packages/backend-common/src/reading/BitbucketUrlReader.ts

```diff
@@ -14,71 +14,249 @@
+        filepathtype !== 'src')
+    ) {
+      throw new Error('Invalid Bitbucket URL or file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+
+    if (provider.host === 'bitbucket.org') {
+      if (!ref) {
+        throw new Error('Invalid Bitbucket URL or file path');
+      }
+      return new URL(
+        `${provider.rawBaseUrl}/repositories/${owner}/${name}/src/${ref}/${pathWithoutSlash}`,
+      );
+    }
+    return new URL(
+      `${provider.rawBaseUrl}/projects/${owner}/repos/${name}/raw/${pathWithoutSlash}?at=${ref}`,
+    );
+  } catch (e) {
+    throw new Error(`Incorrect URL: ${target}, ${e}`);
+  }
+}
+
+export function readConfig(config: Config): ProviderConfig[] {
+  const providers: ProviderConfig[] = [];
 
   const providerConfigs =
     config.getOptionalConfigArray('integrations.bitbucket') ?? [];
 
+  // First read all the explicit providers
   for (const providerConfig of providerConfigs) {
     const host = providerConfig.getOptionalString('host') ?? 'bitbucket.org';
+    let apiBaseUrl = providerConfig.getOptionalString('apiBaseUrl');
+    let rawBaseUrl = providerConfig.getOptionalString('rawBaseUrl');
+    const token = providerConfig.getOptionalString('token');
+    const username = providerConfig.getOptionalString('username');
+    const password = providerConfig.getOptionalString('appPassword');
+
+    if (apiBaseUrl) {
+      apiBaseUrl = apiBaseUrl.replace(/\/+$/, '');
+    } else if (host === 'bitbucket.org') {
+      apiBaseUrl = 'https://api.bitbucket.org/2.0';
+    }
+
+    if (rawBaseUrl) {
+      rawBaseUrl = rawBaseUrl.replace(/\/+$/, '');
+    } else if (host === 'bitbucket.org') {
+      rawBaseUrl = 'https://api.bitbucket.org/2.0';
+    }
 
-    let auth;
-    if (providerConfig.has('username')) {
-      const username = providerConfig.getString('username');
-      const appPassword = providerConfig.getString('appPassword');
-      auth = { username, appPassword };
+    if (!apiBaseUrl && !rawBaseUrl) {
+      throw new Error(
+        `Bitbucket integration for '${host}' must configure an explicit apiBaseUrl and rawBaseUrl`,
+      );
+    }
+    if (!token && username && !password) {
+      throw new Error(
+        `Bitbucket integration for '${host}' has configured a username but is missing a required password.`,
+      );
     }
 
-    optionsArr.push({ host, auth });
+    providers.push({ host, apiBaseUrl, rawBaseUrl, token, username, password });
   }
 
-  // As a convenience we always make sure there's at least an unauthenticated
-  // reader for public bitbucket repos.
-  if (!optionsArr.some(p => p.host === 'bitbucket.org')) {
-    optionsArr.push({ host: 'bitbucket.org' });
+  // If no explicit bitbucket.org provider was added, put one in the list as
+  // a convenience
+  if (!providers.some(p => p.host === 'bitbucket.org')) {
+    providers.push({
+      host: 'bitbucket.org',
+      apiBaseUrl: 'https://api.bitbucket.org/2.0',
```

Comment:
maybe make a `const DEFAULT_BASE_URL` or something out of this, it's repeated a couple times